### ISSUE TITLE
0.41.0: Regression regarding gnome.gtkdoc linker flags

### ISSUE THREAD

**Issue by inigomartinez** [mesonbuild/meson#1946]
Although meson 0.40.[01] does work, in the new 0.41.0 version when using gnome module's gtkdoc function, it doesn't use linker flags properly which ends with the following error:

```
Error in gtkdoc helper script:
'gtkdoc-scangobj' failed with status 127
./program-scan: error while loading shared libraries: libprogram.so.0: cannot open shared object file: No such file or directory
Scan failed: 
```

**Comment by nirbheek** [comment]
We need more details to figure this out. Could you please reduce this to a small test case and attach it here?

**Comment by inigomartinez** [comment]
Although I'm aware that is not a small test, this is what I was working on:

https://github.com/inigomartinez/gnome-todo/tree/meson-update

It generates documentation properly with 0.40.1 while it doesn't with 0.41.0.

First, [I do generate a shared library](https://github.com/inigomartinez/gnome-todo/blob/meson-update/src/meson.build#L186), which is going to be used for documentation generation:

```
libgtd = shared_library(
  'gtd',
  sources: sources,
  version: libversion,
  soversion: soversion,
  include_directories: gnome_todo_incs,
  dependencies: gnome_todo_deps,
  c_args: gnome_todo_cflags,
  link_args: gnome_todo_common_ldflags
)

libgtd_dep = declare_dependency(
  link_with: libgtd,
  include_directories: include_directories('.'),
  dependencies: gnome_todo_deps
)
```

Then using this library dependency, [documentation is generated](https://github.com/inigomartinez/gnome-todo/blob/meson-update/doc/reference/meson.build#L53) in the following way:

```
gnome.gtkdoc(
  meson.project_name(),
  main_sgml: meson.project_name() + '-docs.sgml',
  src_dir: join_paths(meson.source_root(), 'src'),
  dependencies: libgtd_dep,
  scan_args: [
    '--rebuild-types',
    '--ignore-headers=' + ' '.join(private_headers),
  ],
  mkdb_args: [
    '--sgml-mode',
    '--output-format=xml'
  ],
  gobject_typesfile: meson.project_name() + '.types',
  fixxref_args: [
    '--extra-dir=@0@'.format(join_paths(glib_doc_path, 'glib')),
    '--extra-dir=@0@'.format(join_paths(glib_doc_path, 'gmodule')),
    '--extra-dir=@0@'.format(join_paths(glib_doc_path, 'gobject')),
    '--extra-dir=@0@'.format(join_paths(glib_doc_path, 'gio')),
    '--extra-dir=@0@'.format(join_paths(gtk_doc_path, 'gdk')),
    '--extra-dir=@0@'.format(join_paths(gtk_doc_path, 'gtk')),
    '--extra-dir=@0@'.format(join_paths(gtk_doc_path, 'gtksourceview-3.0')),
  ],
  html_assets: html_images,
  install: true,
  install_dir: gnome_todo_doc_path
)
```

Please don't hesitate on asking me any question, or if you want me to make another program, though I will need some time to something new from scratch.

**Comment by nirbheek** [comment]
I finally had time to look at this, but it seems your repository is no more. Could you please upload the testcase somewhere?

**Comment by inigomartinez** [comment]
I'm sorry, I removed my fork recently because changes were accepted in the main repository. You can find the main repository at:

https://git.gnome.org/browse/gnome-todo ([GitHub mirror](https://github.com/GNOME/gnome-todo))

Another applications which also suffers the same problem is devhelp. Althought the meson port has not yet been accepted, there is a branch at:

https://git.gnome.org/browse/devhelp/log/?h=wip/meson ([GitHub mirror](https://github.com/GNOME/devhelp/tree/wip/meson))

Please, don't hesitate to contact me if you have any further questions.

**Comment by nirbheek** [comment]
Okay, I figured out that it's moved to https://git.gnome.org/browse/gnome-todo and you have to pass `-Denable-gtk-doc=true` to reproduce this. There's two problems:

1. The ldflags passed to gtkdoc are `-lgtd -Lsrc -L$builddir/src -Wl,-rpath,src` which is the wrong order (easily fixed)
2. The scan binary `gnome-todo-scan` is run from `docs/reference`, but the rpath assumes it is in the build root, so it can't find libgtd.so at runtime

@ebassi and @matthiasclasen were talking about (2) on IRC a while ago. I'll try to come up with a fix for it.

**Comment by ebassi** [comment]
The problem we were discussing on IRC was caused by inclusion paths used by the gnome.gtkdoc() method, not by linker flags.

The dependencies are resolved at the point of the module, but the generated scan binary is executed from a different work directory, so the inclusions directives are wrong.

**Comment by nirbheek** [comment]
I think the GNOME module needs a bit of cleanup. It reinvents a lot of backend-specific logic, which causes the issue that you were talking about, and also means it doesn't work with the VS backends.

**Comment by nirbheek** [comment]
Actually, turns out this was already fixed by 6bc14424b40b0741112737859b576ed9533f099a. https://github.com/mesonbuild/meson/pull/2072 now contains unrelated fixes. Closing.
